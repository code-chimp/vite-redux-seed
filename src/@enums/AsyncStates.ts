enum AsyncStates {
  Idle = 'IDLE',
  Pending = 'PENDING',
  Success = 'SUCCESS',
  Fail = 'FAIL',
}

export default AsyncStates;
