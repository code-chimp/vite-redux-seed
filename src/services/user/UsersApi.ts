import IUser from '../../@interfaces/IUser';
import { createGetRequest, processApiResponse, unwrapServiceError } from '../../helpers';

export const fetchUsers = async (): Promise<Array<IUser>> => {
  try {
    const response = await fetch(createGetRequest('/users'));

    return await processApiResponse<Array<IUser>>(response);
  } catch (e) {
    throw unwrapServiceError('UsersApi.fetchUsers', e);
  }
};

export const fetchUser = async (id: number): Promise<IUser> => {
  try {
    const response = await fetch(createGetRequest(`/users/${id}`));

    return await processApiResponse<IUser>(response);
  } catch (e) {
    throw unwrapServiceError('UsersApi.fetchUser', e);
  }
};
