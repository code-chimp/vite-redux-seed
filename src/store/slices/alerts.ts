import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 as uuid } from 'uuid';
import AlertTypes from '../../@enums/AlertTypes';
import AlertType from '../../@types/AlertType';
import IAlert from '../../@interfaces/IAlert';
import { IStore } from '../';

const initialState: Array<IAlert> = [];

const createAlertPayload = (type: AlertType, alert: Partial<IAlert>): IAlert => {
  return {
    ...alert,
    id: uuid(),
    type,
  };
};

export const alerts = createSlice({
  name: 'alerts',
  initialState,
  reducers: {
    removeAlert: (state: Array<IAlert>, action: PayloadAction<string>) => {
      const index = state.findIndex(a => a.id === action.payload);
      if (index > -1) {
        state.splice(index, 1);
      }
    },
    addErrorAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Error, alert) };
      },
    },
    addInfoAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Info, alert) };
      },
    },
    addSuccessAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Success, alert) };
      },
    },
    addWarningAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Warning, alert) };
      },
    },
  },
});

export const { addErrorAlert, addInfoAlert, addSuccessAlert, addWarningAlert, removeAlert } =
  alerts.actions;

export const selectAlerts = (state: IStore) => state.alerts;

export default alerts.reducer;
