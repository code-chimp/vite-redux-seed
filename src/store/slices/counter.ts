import { createSlice } from '@reduxjs/toolkit';
import { IStore } from '../';

export interface ICounterSlice {
  value: number;
}

export const counter = createSlice({
  name: 'counter',
  initialState: {
    value: 0,
  },
  reducers: {
    increment: (state: ICounterSlice) => {
      state.value += 1;
    },
    decrement: (state: ICounterSlice) => {
      state.value -= 1;
    },
    incrementByAmount: (state: ICounterSlice, action: { type: string; payload: number }) => {
      state.value += action.payload;
    },
  },
});

/* Ignore selectors in test coverage report due to simplicity */
/* istanbul ignore next */
export const selectCount = (state: IStore) => state.counter.value;

export const { increment, decrement, incrementByAmount } = counter.actions;

export default counter.reducer;
