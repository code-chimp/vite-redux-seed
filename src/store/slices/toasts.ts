import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 as uuid } from 'uuid';
import IToastMessage from '../../@interfaces/IToastMessage';
import ToastTypes from '../../@enums/ToastTypes';
import { IStore } from '../';

const initialState: Array<IToastMessage> = [];

function createToastMessagePayload(toast: Partial<IToastMessage>): IToastMessage {
  return {
    ...toast,
    id: uuid(),
  };
}

export const toasts = createSlice({
  name: 'toasts',
  initialState,
  reducers: {
    removeToastMessage: (state: Array<IToastMessage>, action: PayloadAction<string>) => {
      const index = state.findIndex(t => t.id === action.payload);
      if (index > -1) {
        state.splice(index, 1);
      }
    },
    addErrorToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Error, text }) };
      },
    },
    addInfoToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Info, text }) };
      },
    },
    addSuccessToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Success, text }) };
      },
    },
    addWarningToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Warning, text }) };
      },
    },
  },
});

export const {
  addErrorToastMessage,
  addInfoToastMessage,
  addSuccessToastMessage,
  addWarningToastMessage,
  removeToastMessage,
} = toasts.actions;

export const selectToasts = (state: IStore) => state.toasts;

export default toasts.reducer;
