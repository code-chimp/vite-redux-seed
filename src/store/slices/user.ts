import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import AsyncStates from '../../@enums/AsyncStates';
import IUser from '../../@interfaces/IUser';
import AsyncStatus from '../../@types/AsyncStatus';
import { fetchUsers } from '../../services/user/UsersApi';
import { IStore } from '../';

export interface IUserSlice {
  current: IUser | null;
  users: Array<IUser>;
  status: AsyncStatus;
  error: string | null;
}

export const initialState: IUserSlice = {
  current: null,
  users: [],
  status: AsyncStates.Idle,
  error: null,
};

export const loadUsers = createAsyncThunk(
  'users/fetchUsers',
  async (arg: void, { rejectWithValue }) => {
    try {
      return await fetchUsers();
    } catch (e: any) {
      return rejectWithValue(e.message || 'Internal Server Error');
    }
  },
);

export const user = createSlice({
  name: 'user',
  initialState,
  reducers: {
    clearCurrentUser: (state: IUserSlice) => {
      state.current = null;
    },
    setCurrentUser: (state: IUserSlice, action: { type: string; payload: number }) => {
      const user = state.users.find(x => x.id === action.payload);

      if (user) {
        state.current = user;
        state.error = null;
      } else {
        state.current = null;
        state.error = 'user not found';
      }
    },
  },
  extraReducers: builder => {
    builder.addCase(loadUsers.pending, (state: IUserSlice) => {
      state.status = AsyncStates.Pending;
    });

    builder.addCase(
      loadUsers.fulfilled,
      (state: IUserSlice, { payload }: PayloadAction<Array<IUser> | undefined>) => {
        state.status = AsyncStates.Success;
        state.users = payload ? (payload as Array<IUser>) : [];
        // or you could make it an additive operation
        // state.users = [...state.users, ...(action.payload as Array<IUser>)];
      },
    );

    builder.addCase(loadUsers.rejected, (state: IUserSlice, action: any) => {
      state.status = AsyncStates.Fail;
      state.error = action.payload as string;
      state.users = [];
    });
  },
});

/* Ignore selectors in test coverage report due to simplicity */
/* istanbul ignore next */
export const selectCurrentUser = (state: IStore) => state.user.current;

/* istanbul ignore next */
export const selectUsers = (state: IStore) => state.user.users;

export const { setCurrentUser, clearCurrentUser } = user.actions;

export default user.reducer;
