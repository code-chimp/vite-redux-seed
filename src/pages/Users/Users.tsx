import React, { FC, useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../../helpers';
import AsyncStates from '../../@enums/AsyncStates';
import IUser from '../../@interfaces/IUser';
import { loadUsers, selectUsers } from '../../store/slices/user';

const Users: FC = () => {
  const dispatch = useAppDispatch();
  const users = useAppSelector(selectUsers);
  // ad-hoc selectors
  const status = useAppSelector(state => state.user.status);
  const error = useAppSelector(state => state.user.error);

  useEffect(() => {
    // Let us suppose that fetching the users is an expensive operation
    // that we want to repeat as little as possible. By keeping the collection
    // in the global store we can reduce the number of queries we will need to
    // make to the back end to retrieve identical information.
    //
    // Should we unload, then later reload this component, the slice will likely
    // have a status of 'AsyncStates.Success' and the users array will contain
    // elements - therefore no "dispatch" to loadUsers will occur
    if (status === AsyncStates.Idle && !users.length) {
      dispatch(loadUsers());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <table className="table">
      <thead>
        <tr className="table-info">
          <th>Name</th>
          <th>City</th>
          <th>Zip</th>
          <th>Company</th>
        </tr>
      </thead>
      <tbody>
        {users.length ? (
          users.map((user: IUser) => (
            <tr key={user.id} data-testid={`row_${user.id}`}>
              <td>{user.name}</td>
              <td>{user.address?.city || 'unknown'}</td>
              <td>{user.address?.zipcode || 'unknown'}</td>
              <td>{user.company?.name || 'unknown'}</td>
            </tr>
          ))
        ) : status === AsyncStates.Fail ? (
          <tr>
            <td data-testid="error-indicator" colSpan={4}>
              Error: {error}
            </td>
          </tr>
        ) : (
          <tr>
            <td data-testid="loading-indicator" colSpan={4}>
              Loading...
            </td>
          </tr>
        )}
      </tbody>
    </table>
  );
};

export default Users;
