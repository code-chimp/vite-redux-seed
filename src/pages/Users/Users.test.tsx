import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore, { MockStore } from 'redux-mock-store';
import Users from './Users';
import AsyncStates from '../../@enums/AsyncStates';
import { initialState, loadUsers } from '../../store/slices/user';

const mockStore = configureMockStore([thunk]);
let store: MockStore;

describe('components / Users', () => {
  it('should match the snapshot', () => {
    store = mockStore({
      user: { ...initialState },
    });

    const { asFragment } = render(
      <Provider store={store}>
        <Users />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a loading message in the default state and dispatch the load action', () => {
    store = mockStore({
      user: { ...initialState },
    });

    render(
      <Provider store={store}>
        <Users />
      </Provider>,
    );

    const loading = screen.getByTestId('loading-indicator');
    expect(loading).toBeInTheDocument();

    const actionsDispatched = store.getActions().map(action => action.type);
    expect(actionsDispatched).toEqual([loadUsers.pending.type]);
  });

  it('should render a loading message in the pending state and not dispatch the load action', () => {
    store = mockStore({
      user: { ...initialState, status: AsyncStates.Pending },
    });

    render(
      <Provider store={store}>
        <Users />
      </Provider>,
    );

    const loading = screen.getByTestId('loading-indicator');
    expect(loading).toBeInTheDocument();

    // does not satisfy the condition inside of the `useEffect`
    expect(store.getActions()).toEqual([]);
  });

  it('should render an error in the failed state', () => {
    store = mockStore({
      user: {
        ...initialState,
        status: AsyncStates.Fail,
        error: 'shabubu',
      },
    });

    render(
      <Provider store={store}>
        <Users />
      </Provider>,
    );

    const error = screen.getByTestId('error-indicator');
    expect(error).toBeInTheDocument();
    expect(error.textContent).toBe('Error: shabubu');

    // should not have satisfied the condition inside of the `useEffect`
    expect(store.getActions()).toEqual([]);
  });

  it('should render users upon success', () => {
    store = mockStore({
      user: {
        ...initialState,
        users: [
          {
            id: 1,
            name: 'A User',
            address: { city: 'Test', zipcode: 'test' },
            company: { name: 'kiewit' },
          },
          {
            id: 2,
            name: 'B User',
            address: { city: 'Test', zipcode: 'test' },
            company: { name: 'kiewit' },
          },
        ],
        status: AsyncStates.Success,
      },
    });

    render(
      <Provider store={store}>
        <Users />
      </Provider>,
    );

    const row1 = screen.getByTestId('row_1');
    expect(row1).toBeInTheDocument();
    const row2 = screen.getByTestId('row_2');
    expect(row2).toBeInTheDocument();

    // should not have satisfied the condition inside of the `useEffect`
    expect(store.getActions()).toEqual([]);
  });
});
