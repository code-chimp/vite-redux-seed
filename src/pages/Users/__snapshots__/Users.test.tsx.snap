// Vitest Snapshot v1

exports[`components / Users > should match the snapshot 1`] = `
<DocumentFragment>
  <table
    class="table"
  >
    <thead>
      <tr
        class="table-info"
      >
        <th>
          Name
        </th>
        <th>
          City
        </th>
        <th>
          Zip
        </th>
        <th>
          Company
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colspan="4"
          data-testid="loading-indicator"
        >
          Loading...
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
